# Importando librerías

In [133]:
from zipfile import ZipFile
import matplotlib as plt
import pandas as pd
import plotly.graph_objects as go
import plotly.express as px

# Consiguiendo los datos

In [71]:
# Leemos nuestro archivo zip 
steam_zip = ZipFile('SteamDataset.zip')

### Construir DataFrame

In [72]:
# Abrimos el archivo csv que contiene el archivo zip para construir nuestro DataFrame
df = pd.read_csv(steam_zip.open('steam.csv'), index_col=[0])

# Limpieza y manipulación de datos

In [73]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 27103 entries, 0 to 27102
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   name                27103 non-null  object 
 1   release_date        27103 non-null  object 
 2   developer           27102 non-null  object 
 3   achievements        27103 non-null  int64  
 4   positive_ratings    27103 non-null  int64  
 5   negative_ratings    27103 non-null  int64  
 6   avg_hours_per_user  27103 non-null  int64  
 7   price               27103 non-null  object 
 8   sells               27103 non-null  float64
 9   action              27103 non-null  bool   
 10  adventure           27103 non-null  bool   
 11  rpg                 27103 non-null  bool   
 12  simulation          27103 non-null  bool   
 13  strategy            27103 non-null  bool   
 14  racing              27103 non-null  bool   
 15  free_to_play        27103 non-null  bool   
 16  early_acc

In [74]:
df.columns

Index(['name', 'release_date', 'developer', 'achievements', 'positive_ratings',
       'negative_ratings', 'avg_hours_per_user', 'price', 'sells', 'action',
       'adventure', 'rpg', 'simulation', 'strategy', 'racing', 'free_to_play',
       'early_access'],
      dtype='object')

In [75]:
df.head()

name release_date         developer  achievements  \
0             Counter-Strike   2000-11-01             Valve             0   
1      Team Fortress Classic   1999-04-01             Valve             0   
2              Day of Defeat   2003-05-01             Valve             0   
3         Deathmatch Classic   2001-06-01             Valve             0   
4  Half-Life: Opposing Force   1999-11-01  Gearbox Software             0   

   positive_ratings  negative_ratings  avg_hours_per_user  price       sells  \
0            124534              3339               17612  7.19€  15000000.0   
1              3318               633                 277  3.99€   7500000.0   
2              3416               398                 187  3.99€   7500000.0   
3              1273               267                 258  3.99€   7500000.0   
4              5250               288                 624  3.99€   7500000.0   

   action  adventure    rpg  simulation  strategy  racing  free_to_play  \
0    True      False  False       False     False   False         False   
1    True      False  False       False     False   False         False   
2    True      False  False       False     False   False         False   
3    True      False  False       False     False   False         False   
4    True      False  False       False     False   False         False   

   early_access  
0         False  
1         False  
2         False  
3         False  
4         False

Lo primero que notaremos es que no se pueden ver todas las columnas de una manera optima, por otro lado, vemos que se utilizan mascaras booleanas para los géneros de cada videojuego, rompiendo con una regla de normalización y teniendo todo bastante desordenado.

**Para solucionar esto, vamos a crear una nueva columna que se llamara "géneros" donde vamos a asignar cada videojuego con su género correspondiente en base a las mascaras booleanas.**

In [76]:
genres = df[['action','adventure','rpg','simulation','strategy','racing']]
# Como un videojuego puede tener más de un género, usaremos el método 'join' para mostrarlos en valores separados por coma.
df['gender']= df.apply(lambda row: ', '.join([col for col in genres.columns if row[col]]),axis=1)

Ahora eliminamos las columnas que sobraron

In [77]:
df.drop(['action','adventure','rpg','simulation','strategy','racing'],axis=1,inplace=True)

In [78]:
df

name release_date  \
0                       Counter-Strike   2000-11-01   
1                Team Fortress Classic   1999-04-01   
2                        Day of Defeat   2003-05-01   
3                   Deathmatch Classic   2001-06-01   
4            Half-Life: Opposing Force   1999-11-01   
...                                ...          ...   
27098  Google Spotlight Stories: Piggy   2018-06-06   
27099              Sausage Sports Club   2018-07-19   
27100                      Party Panic   2017-08-07   
27101                         SoulHunt   2016-12-23   
27102                        Quickshot   2018-03-12   

                       developer  achievements  positive_ratings  \
0                          Valve             0            124534   
1                          Valve             0              3318   
2                          Valve             0              3416   
3                          Valve             0              1273   
4               Gearbox Software             0              5250   
...                          ...           ...               ...   
27098                Google Inc.             0                15   
27099                 Chris Wade            18                16   
27100  Everglow Interactive Inc.            28               619   
27101         WinterCell Studios            18                56   
27102               Headroom.one             0                 2   

       negative_ratings  avg_hours_per_user   price       sells  free_to_play  \
0                  3339               17612   7.19€  15000000.0         False   
1                   633                 277   3.99€   7500000.0         False   
2                   398                 187   3.99€   7500000.0         False   
3                   267                 258   3.99€   7500000.0         False   
4                   288                 624   3.99€   7500000.0         False   
...                 ...                 ...     ...         ...           ...   
27098                 0                   0    0.0€     10000.0          True   
27099                 3                   0  10.99€     10000.0         False   
27100               226                  84  12.99€     75000.0         False   
27101                19                   0   5.99€     10000.0         False   
27102                 1                   0   2.09€     10000.0         False   

       early_access             gender  
0             False             action  
1             False             action  
2             False             action  
3             False             action  
4             False             action  
...             ...                ...  
27098         False          adventure  
27099         False  action, adventure  
27100         False             action  
27101          True             action  
27102         False             action  

[27103 rows x 12 columns]

In [79]:
df[df.gender == '']

name release_date  \
26                         Rag Doll Kung Fu   2005-10-12   
33                               Multiwinia   2008-09-19   
41                         Source Filmmaker   2012-07-10   
62               Gumboy - Crazy Adventures™   2006-12-19   
90                              Zuma Deluxe   2006-08-30   
...                                     ...          ...   
27085                     Whispering Flames   2018-11-17   
27087                    Power Solitaire VR   2017-06-07   
27088                        MiniGolf Mania   2016-03-25   
27093           Viki Spotter: Space Mission   2018-05-11   
27094  Champ Against Chumps Upgrade Edition   2018-04-20   

                     developer  achievements  positive_ratings  \
26                 Mark Healey             0                40   
33       Introversion Software            12               368   
41                       Valve             0             15083   
62             CINEMAX, s.r.o.             0                52   
90          PopCap Games, Inc.             0               437   
...                        ...           ...               ...   
27085               Red Cattus            10                 9   
27087  Parkerhill Reality Labs             0                17   
27088        Road Turtle Games             0                36   
27093                 For Kids           279                70   
27094              MisterTtocS            22                 2   

       negative_ratings  avg_hours_per_user  price      sells  free_to_play  \
26                   17                   0  5.99€    35000.0         False   
33                   75                 121  7.19€   750000.0         False   
41                 1111                 239   0.0€  1500000.0         False   
62                   46                   0   0.0€    75000.0         False   
90                   37                 267  4.25€   150000.0         False   
...                 ...                 ...    ...        ...           ...   
27085                 4                   0  2.09€    10000.0         False   
27087                 1                   0   0.0€    10000.0          True   
27088                16                   0   0.0€    10000.0         False   
27093                 1                   0  1.69€    10000.0         False   
27094                 0                   0  1.69€    10000.0         False   

       early_access gender  
26            False         
33            False         
41            False         
62            False         
90            False         
...             ...    ...  
27085         False         
27087         False         
27088         False         
27093         False         
27094         False         

[3716 rows x 12 columns]

Oh... podemos notar que hubo varios videojuegos que han quedado sin género. ¿Qué podemos hacer en este caso? 
Bueno, insertar su género correspondiente uno por uno no es una opción. Por lo que voy a insertar "None" para todos aquellos videojuegos que han quedado sin género. 

In [80]:
df['gender'] = df['gender'].replace('','none')

In [81]:
df[df.gender == ''].value_counts()

Series([], Name: count, dtype: int64)

Tendremos que definir a qué género pertenece un juego, porque hay juegos que pertenecen a varios géneros. Entonces, ¿qué se hace en este caso? 
Tendremos que tomar la decisión de que el primer género es el que mejor define el juego.

In [82]:
pat_genres = ',.*'

genres = df['gender'].str.replace(pat_genres, '', regex=True)
df['gender'] = genres

In [83]:
df

name release_date  \
0                       Counter-Strike   2000-11-01   
1                Team Fortress Classic   1999-04-01   
2                        Day of Defeat   2003-05-01   
3                   Deathmatch Classic   2001-06-01   
4            Half-Life: Opposing Force   1999-11-01   
...                                ...          ...   
27098  Google Spotlight Stories: Piggy   2018-06-06   
27099              Sausage Sports Club   2018-07-19   
27100                      Party Panic   2017-08-07   
27101                         SoulHunt   2016-12-23   
27102                        Quickshot   2018-03-12   

                       developer  achievements  positive_ratings  \
0                          Valve             0            124534   
1                          Valve             0              3318   
2                          Valve             0              3416   
3                          Valve             0              1273   
4               Gearbox Software             0              5250   
...                          ...           ...               ...   
27098                Google Inc.             0                15   
27099                 Chris Wade            18                16   
27100  Everglow Interactive Inc.            28               619   
27101         WinterCell Studios            18                56   
27102               Headroom.one             0                 2   

       negative_ratings  avg_hours_per_user   price       sells  free_to_play  \
0                  3339               17612   7.19€  15000000.0         False   
1                   633                 277   3.99€   7500000.0         False   
2                   398                 187   3.99€   7500000.0         False   
3                   267                 258   3.99€   7500000.0         False   
4                   288                 624   3.99€   7500000.0         False   
...                 ...                 ...     ...         ...           ...   
27098                 0                   0    0.0€     10000.0          True   
27099                 3                   0  10.99€     10000.0         False   
27100               226                  84  12.99€     75000.0         False   
27101                19                   0   5.99€     10000.0         False   
27102                 1                   0   2.09€     10000.0         False   

       early_access     gender  
0             False     action  
1             False     action  
2             False     action  
3             False     action  
4             False     action  
...             ...        ...  
27098         False  adventure  
27099         False     action  
27100         False     action  
27101          True     action  
27102         False     action  

[27103 rows x 12 columns]

## Quitar datos duplicados o irrelevantes

Vamos a contar directamente las columnas donde hayan valores duplicados

In [84]:
df.duplicated().any()

np.True_

In [85]:
df.duplicated().unique()

array([False,  True])

In [86]:
df.duplicated().value_counts()

False    27075
True        28
Name: count, dtype: int64

27075 filas son únicas y 28 de ellas contienen valores duplicados, vamos a eliminarlas!

In [87]:
df.drop_duplicates(inplace=True)

In [88]:
df.duplicated().value_counts()

False    27075
Name: count, dtype: int64

## Arreglar errores estructurales

In [89]:
print(df['release_date'].dtype)

object


Las fechas son strings, vamos a convertirlas a un formato adecuado usando 'pd.to_datetime'

In [90]:
df['release_date'] = pd.to_datetime(df['release_date'])

Otra cosa que notamos es que los precios no están en un formato adecuado, al momento de tener el simbolo de la moneda "€" ya impide realizar operaciones. Por lo que hay que solucionarlo.

In [91]:
df['price'] = df['price'].str.replace('€','').apply(float)

## Filtrar outliers 

In [92]:
df.describe()

release_date  achievements  positive_ratings  \
count                          27075  27075.000000      2.707500e+04   
mean   2016-12-31 14:21:17.252077568     45.248864      1.000559e+03   
min              1997-06-30 00:00:00      0.000000      0.000000e+00   
25%              2016-04-04 00:00:00      0.000000      6.000000e+00   
50%              2017-08-08 00:00:00      7.000000      2.400000e+01   
75%              2018-06-06 12:00:00     23.000000      1.260000e+02   
max              2019-05-01 00:00:00   9821.000000      2.644404e+06   
std                              NaN    352.670281      1.898872e+04   

       negative_ratings  avg_hours_per_user         price         sells  
count      27075.000000        27075.000000  27075.000000  2.707500e+04  
mean         211.027147          149.804949      6.078193  1.340905e+05  
min            0.000000            0.000000      0.000000  1.000000e+04  
25%            2.000000            0.000000      1.690000  1.000000e+04  
50%            9.000000            0.000000      3.990000  1.000000e+04  
75%           42.000000            0.000000      7.190000  3.500000e+04  
max       487076.000000       190625.000000    421.990000  1.500000e+08  
std         4284.938531         1827.038141      7.874922  1.328089e+06

No es necesario eliminar los outliers, por el momento, aplicaremos otras técnicas para que no afecten la conclusión final (si es necesario)

## Lidiar con NANs

In [93]:
df.isna()

name  release_date  developer  achievements  positive_ratings  \
0      False         False      False         False             False   
1      False         False      False         False             False   
2      False         False      False         False             False   
3      False         False      False         False             False   
4      False         False      False         False             False   
...      ...           ...        ...           ...               ...   
27070  False         False      False         False             False   
27071  False         False      False         False             False   
27072  False         False      False         False             False   
27073  False         False      False         False             False   
27074  False         False      False         False             False   

       negative_ratings  avg_hours_per_user  price  sells  free_to_play  \
0                 False               False  False  False         False   
1                 False               False  False  False         False   
2                 False               False  False  False         False   
3                 False               False  False  False         False   
4                 False               False  False  False         False   
...                 ...                 ...    ...    ...           ...   
27070             False               False  False  False         False   
27071             False               False  False  False         False   
27072             False               False  False  False         False   
27073             False               False  False  False         False   
27074             False               False  False  False         False   

       early_access  gender  
0             False   False  
1             False   False  
2             False   False  
3             False   False  
4             False   False  
...             ...     ...  
27070         False   False  
27071         False   False  
27072         False   False  
27073         False   False  
27074         False   False  

[27075 rows x 12 columns]

In [94]:
df.isna().any()

name                  False
release_date          False
developer              True
achievements          False
positive_ratings      False
negative_ratings      False
avg_hours_per_user    False
price                 False
sells                 False
free_to_play          False
early_access          False
gender                False
dtype: bool

In [95]:
df[df["developer"].isna()]

name release_date developer  achievements  \
23071  The Battle Of Ages   2018-09-07       NaN             0   

       positive_ratings  negative_ratings  avg_hours_per_user  price    sells  \
23071                 2                 1                   0   3.99  10000.0   

       free_to_play  early_access  gender  
23071         False          True  action

En la página oficial de Steam, la desarrolladora no aparece en la información del juego. Por lo que este valor está justificado.

## Validar nuestros datos

In [96]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 27075 entries, 0 to 27074
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   name                27075 non-null  object        
 1   release_date        27075 non-null  datetime64[ns]
 2   developer           27074 non-null  object        
 3   achievements        27075 non-null  int64         
 4   positive_ratings    27075 non-null  int64         
 5   negative_ratings    27075 non-null  int64         
 6   avg_hours_per_user  27075 non-null  int64         
 7   price               27075 non-null  float64       
 8   sells               27075 non-null  float64       
 9   free_to_play        27075 non-null  bool          
 10  early_access        27075 non-null  bool          
 11  gender              27075 non-null  object        
dtypes: bool(2), datetime64[ns](1), float64(2), int64(4), object(3)
memory usage: 2.3+ MB


In [97]:
df[["achievements","positive_ratings","negative_ratings","avg_hours_per_user","price","sells"]].corr().style.background_gradient()

# EDA

## Cual es el TOP10 de juegos más jugados?

In [98]:
df.sort_values(by="avg_hours_per_user",ascending=False).head(10)
mas_jugados = df.sort_values(by="avg_hours_per_user",ascending=False).head(10)

In [99]:
fig = px.bar(x=mas_jugados["name"],
             y=mas_jugados["avg_hours_per_user"],
             color=mas_jugados.index,
             opacity=0.9, 
             color_discrete_sequence=px.colors.qualitative.Set2)

fig.update_layout(title='Top 10 Videojuegos más jugados',
                  font=dict(family='Verdana', size=16, color='black'),
                  template='simple_white',
                  height=500,
                  width=900,
                  legend_title_text="Ranking",
                  xaxis_title="Nombre",
                  yaxis_title="Promedio de horas jugadas",
                  xaxis=dict(showgrid=False, tickangle=45),
                  yaxis=dict(showgrid=True, gridcolor="lightgrey"))

fig.show()


## Cuál es el género de videojuegos más vendido?

In [100]:
df.groupby('gender')['sells'].sum().max()

np.float64(2361690000.0)

In [101]:
genre_sales = df.groupby('gender')['sells'].sum().sort_values(ascending=False)[:5]

In [102]:
genre_sales

gender
action        2.361690e+09
adventure     3.713400e+08
simulation    2.645850e+08
strategy      2.183800e+08
rpg           2.169100e+08
Name: sells, dtype: float64

In [103]:
fig = px.bar(x=genre_sales.index,
             y=genre_sales,
             color=genre_sales.index,
             opacity=0.9, 
             color_discrete_sequence=px.colors.qualitative.Set2)

fig.update_layout(title='Géneros de videojuegos más vendidos',
                  font=dict(family='Verdana', size=16, color='black'),
                  template='simple_white',
                  height=500,
                  width=900,
                  legend_title_text="Géneros",
                  xaxis_title="Géneros",
                  yaxis_title="Ventas de juegos",
                  xaxis=dict(showgrid=False, tickangle=45),
                  yaxis=dict(showgrid=True, gridcolor="lightgrey"))

fig.show()


## Qué géneros tienen mejores reviews de media?

In [104]:
top_reviews = df.groupby('gender')['positive_ratings'].mean().apply(int).sort_values(ascending=False)[:10]
top_reviews


gender
action        1581
rpg           1367
simulation     820
strategy       566
adventure      430
racing         411
none           228
Name: positive_ratings, dtype: int64

In [105]:
fig = px.bar(x=top_reviews.index,
             y=top_reviews,
             color=top_reviews.index,
             opacity=0.9, 
             color_discrete_sequence=px.colors.qualitative.Set1)

fig.update_layout(title='Videojuegos con más reseñas de media',
                  font=dict(family='Verdana', size=16, color='black'),
                  template='simple_white',
                  height=500,
                  width=900,
                  legend_title_text="Videojuegos",
                  xaxis_title="Videojuegos",
                  yaxis_title="Promedio de reseñas",
                  xaxis=dict(showgrid=False, tickangle=45),
                  yaxis=dict(showgrid=True, gridcolor="lightgrey"))

fig.show()


## Cuánto ha ganado el desarrollador que más juegos ha vendido?

In [106]:
developer_sales = df.groupby('developer')['sells'].sum().apply(int).sort_values(ascending=False)[:5]
developer_sales

developer
Valve                              405750000
PUBG Corporation                    75000000
Valve;Hidden Path Entertainment     75000000
Bethesda Game Studios               37050000
Digital Extremes                    36575000
Name: sells, dtype: int64

In [107]:
i_dev = developer_sales.index
i_dev

Index(['Valve', 'PUBG Corporation', 'Valve;Hidden Path Entertainment',
       'Bethesda Game Studios', 'Digital Extremes'],
      dtype='object', name='developer')

In [108]:
df_dev = df[df['developer'].isin(i_dev)]
df_dev.head(5)

name release_date developer  achievements  \
0         Counter-Strike   2000-11-01     Valve             0   
1  Team Fortress Classic   1999-04-01     Valve             0   
2          Day of Defeat   2003-05-01     Valve             0   
3     Deathmatch Classic   2001-06-01     Valve             0   
5               Ricochet   2000-11-01     Valve             0   

   positive_ratings  negative_ratings  avg_hours_per_user  price       sells  \
0            124534              3339               17612   7.19  15000000.0   
1              3318               633                 277   3.99   7500000.0   
2              3416               398                 187   3.99   7500000.0   
3              1273               267                 258   3.99   7500000.0   
5              2758               684                 175   3.99   7500000.0   

   free_to_play  early_access  gender  
0         False         False  action  
1         False         False  action  
2         False         False  action  
3         False         False  action  
5         False         False  action

In [109]:
avg_price = df_dev['price'].mean()
avg_price

np.float64(8.836585365853658)

In [110]:
total_sales = developer_sales[i_dev] * avg_price
total_sales

developer
Valve                              3.585445e+09
PUBG Corporation                   6.627439e+08
Valve;Hidden Path Entertainment    6.627439e+08
Bethesda Game Studios              3.273955e+08
Digital Extremes                   3.231981e+08
Name: sells, dtype: float64

In [111]:
fig = px.bar(x=total_sales.index,
             y=total_sales,
             color=total_sales.index,
             opacity=0.8, 
             color_discrete_sequence=px.colors.qualitative.Set2) 

fig.update_layout(title='Ganancias de las desarrolladoras en base a sus ventas',
                  font=dict(family='Verdana', size=16, color='black'), 
                  template='simple_white', 
                  height=500,
                  width=900,
                  legend_title_text="Desarrolladoras", 
                  xaxis_title="Géneros",
                  yaxis_title="Ventas de juegos",
                  xaxis=dict(showgrid=False, tickangle=45),
                  yaxis=dict(showgrid=True, gridcolor="lightgrey"))

fig.show()


## Cómo afecta a las reviews que un juego sea Early Access?

In [112]:
early = df[df.early_access == True]
no_early = df[df.early_access == False]

In [113]:
early_pos, early_neg = early["positive_ratings"].sum(), early["negative_ratings"].sum()
not_early_pos, not_early_neg = no_early["positive_ratings"].sum(),no_early["negative_ratings"].sum()

In [114]:
total_early = early_pos + early_neg
total_not_early = not_early_pos + not_early_neg

early_pos_pct,early_neg_pct = (early_pos / total_early) * 100, (early_neg / total_early) * 100
not_early_pos_pct, not_early_neg_pct = (not_early_pos / total_not_early) * 100, (not_early_neg / total_not_early) * 100
  

In [115]:
df_reviews = pd.DataFrame({
    "Type": ["Early Access", "No Early Access"],
    "Positive": [early_pos_pct, not_early_pos_pct],
    "Negative": [early_neg_pct, not_early_neg_pct]
})
df_reviews

Type   Positive   Negative
0     Early Access  76.125303  23.874697
1  No Early Access  82.893239  17.106761

In [116]:
fig = go.Figure()

fig.add_trace(go.Pie(
    labels=["Positivas", "Negativas"],
    values=[early_pos_pct, early_neg_pct],
    name="Early Access",
    hole=0.4,
    marker=dict(colors=["green", "red"])
))

fig.add_trace(go.Pie(
    labels=["Positivas", "Negativas"],
    values=[not_early_pos_pct, not_early_neg_pct],
    name="No Early Access",
    hole=0.4,
    marker=dict(colors=["green", "red"]),
    domain={"x": [0.55, 1]} 
))

fig.update_layout(
    title_text="Proporción de Reviews en Early Access vs No Early Access",
    grid={"rows": 1, "columns": 2}
)

fig.show()

## Cuál es el desarrollador (o desarrolladores) que más tipos de juego y cantidad ha desarrollado?

In [117]:
developer_stats = df.groupby('developer').agg(
    num_genres=('gender', 'nunique'),  
    num_games=('name', 'count') 
)
developer_stats = developer_stats.sort_values(by=['num_genres','num_games'], ascending=False)[:10]
developer_stats

num_genres  num_games
developer                                        
Laush Dmitriy Sergeevich             6         51
Dexion Games                         6         45
RewindApp                            6         43
Snkl Studio                          6         21
KOEI TECMO GAMES CO., LTD.           5         72
Nikita "Ghost_RUS"                   5         50
EnsenaSoft                           5         35
Winter Wolves                        5         24
CSM                                  5         21
VIS-Games                            5         19

In [118]:
fig = px.bar(developer_stats, 
             x=developer_stats.index, 
             y=['num_genres', 'num_games'], 
             title='Desarrolladores con más géneros y juegos',
             labels={'num_genres': 'Número de Géneros', 'num_games': 'Número de Juegos'},
             height=500,
             width=900)

fig.update_layout(
    barmode='stack',  
    font=dict(family='Verdana', size=16, color='black'),
    xaxis_title="Desarrollador", 
    yaxis_title="Cantidad",
    legend_title="Categoría"
)

fig.show()

## En qué mes, desde que tenemos datos, se han publicado más juegos?

In [119]:
df['release_month'] = df.release_date.dt.month
df['release_month'].value_counts().get(11,0)

np.int64(2328)

In [120]:
games_per_month = df.groupby('release_month')['name'].count()
games_per_month.columns = ['name','count']
games_per_month


release_month
1     1974
2     2325
3     2583
4     2437
5     2060
6     1767
7     2186
8     2333
9     2370
10    2491
11    2328
12    2221
Name: name, dtype: int64

In [121]:
fig = px.bar(x=games_per_month.index,
             y=games_per_month.values,
             title='Juegos publicados por mes',
             labels={'x': 'Mes', 'y': 'Número de Juegos'},
             color=games_per_month.index,
             color_continuous_scale='Viridis',
             text=games_per_month.values)

# Personalizar el gráfico
fig.update_layout(
    xaxis=dict(tickmode='array', tickvals=list(range(1, 13)), ticktext=['Ene', 'Feb', 'Mar', 'Abr', 'May', 'Jun', 'Jul', 'Ago', 'Sep', 'Oct', 'Nov', 'Dic']),
    title_font=dict(family='verdana', size=16, color='black'),
    font=dict(family='verdana', size=14),
    template='plotly_white',
    height=500,
    width=900
)

fig.show()

## Cuántas horas se juega de media a cada tipo de juego mensualmente?

In [122]:
monthly_hours_by_genre = df.groupby(['gender', 'release_month'])['avg_hours_per_user'].sum().reset_index()
games_count = df.groupby(['gender', 'release_month']).size().reset_index(name='count')
result = pd.merge(monthly_hours_by_genre, games_count, on=['gender', 'release_month'])
result['avg_hours_per_game'] = result['avg_hours_per_user'] / result['count']
result


gender  release_month  avg_hours_per_user  count  avg_hours_per_game
0     action              1               80069    864           92.672454
1     action              2              130406   1013          128.732478
2     action              3              147500   1144          128.933566
3     action              4              139716   1050          133.062857
4     action              5              159430    958          166.419624
..       ...            ...                 ...    ...                 ...
79  strategy              8               29290    173          169.306358
80  strategy              9               28701    158          181.651899
81  strategy             10               32751    168          194.946429
82  strategy             11               50997    128          398.414062
83  strategy             12               20377    158          128.968354

[84 rows x 5 columns]

In [123]:
fig = px.bar(result, 
             x='release_month', 
             y='avg_hours_per_game', 
             color='gender', 
             barmode='group',
             title='¿Cuántas horas se juega de media a cada tipo de juego mensualmente?',
             labels={'release_month': 'Mes', 'avg_hours_per_game': 'Promedio de Horas por Juego'},
             color_continuous_scale='Viridis',
             text='avg_hours_per_game')

fig.update_layout(
    xaxis=dict(tickmode='array', tickvals=list(range(1, 13)), ticktext=['Ene', 'Feb', 'Mar', 'Abr', 'May', 'Jun', 'Jul', 'Ago', 'Sep', 'Oct', 'Nov', 'Dic']),
    title_font=dict(family='verdana', size=16, color='black'),
    font=dict(family='verdana', size=14),
    template='plotly_white',
    height=500,
    width=900
)

fig.show()

## Los juegos Free to Play tienen mejores o peores reviews?

In [124]:

f2p_games = df[df['free_to_play'] == True]
not_f2p_games = df[df['free_to_play'] == False]

f2p_positive, f2p_negative = f2p_games['positive_ratings'].sum(), f2p_games['negative_ratings'].sum()
not_f2p_positive, not_f2p_negative = not_f2p_games['positive_ratings'].sum(), not_f2p_games['negative_ratings'].sum()

f2p_total = f2p_positive + f2p_negative
not_f2p_total = not_f2p_positive + not_f2p_negative

f2p_pos_pct, f2p_neg_pct = (f2p_positive / f2p_total) * 100, (f2p_negative / f2p_total) * 100
not_f2p_pos_pct, not_f2p_neg_pct = (not_f2p_positive / not_f2p_total) * 100, (not_f2p_negative / not_f2p_total) * 100


In [125]:
df_f2p = pd.DataFrame({
    "Type": ["Free To Play", "No Free To Play"],
    "Positive": [f2p_pos_pct, not_f2p_pos_pct],
    "Negative": [f2p_neg_pct, not_f2p_neg_pct]
})
df_f2p

Type   Positive   Negative
0     Free To Play  82.833235  17.166765
1  No Free To Play  82.488633  17.511367

In [126]:
fig = go.Figure()

fig.add_trace(go.Pie(
    labels=["Positivas", "Negativas"],
    values=[f2p_pos_pct, f2p_neg_pct],
    name="Free to Play",
    hole=0.4,
    marker=dict(colors=["green", "red"])
))

fig.add_trace(go.Pie(
    labels=["Positivas", "Negativas"],
    values=[not_f2p_pos_pct, not_f2p_neg_pct],
    name="No F2P",
    hole=0.4,
    marker=dict(colors=["green", "red"]),
    domain={"x": [0.55, 1]} 
))

fig.update_layout(
    title_text="Proporción de Reviews en juegos Free to Play vs No F2P",
    grid={"rows": 1, "columns": 2}
)

fig.show()

## Cómo ha evolucionado el número de achievements (trofeos) por juego?

In [127]:
df_achievements = df[df['achievements'] > 0]
df_achievements

name release_date         developer  achievements  \
9                   Half-Life 2   2004-11-16             Valve            33   
10       Counter-Strike: Source   2004-11-01             Valve           147   
12        Day of Defeat: Source   2010-07-12             Valve            54   
16     Half-Life 2: Episode One   2006-06-01             Valve            13   
17                       Portal   2007-10-10             Valve            15   
...                         ...          ...               ...           ...   
27054              Magic Clouds   2019-04-19          Wloop777            15   
27058           Nyasha Valkyrie   2019-04-19  Zloy Krot Studio            32   
27069                 Deatherem   2019-04-23             Anima             1   
27070           Room of Pandora   2019-04-24       SHEN JIAWEI             7   
27072          Super Star Blast   2019-04-24       EntwicklerX            24   

       positive_ratings  negative_ratings  avg_hours_per_user  price  \
9                 67902              2419                 691   7.19   
10                76640              3497                6842   7.19   
12                10489              1210                1356   7.19   
16                 7908               517                 281   5.79   
17                51801              1080                 288   7.19   
...                 ...               ...                 ...    ...   
27054                 3                 0                   0   0.79   
27058                 7                 0                   0   0.79   
27069                 2                 2                   0   2.09   
27070                 3                 0                   0   2.09   
27072                 0                 1                   0   3.99   

            sells  free_to_play  early_access     gender  release_month  
9      15000000.0         False         False     action             11  
10     15000000.0         False         False     action             11  
12      7500000.0         False         False     action              7  
16      7500000.0         False         False     action              6  
17     15000000.0         False         False     action             10  
...           ...           ...           ...        ...            ...  
27054     10000.0         False         False  adventure              4  
27058     10000.0         False         False       none              4  
27069     10000.0         False         False       none              4  
27070     10000.0         False         False  adventure              4  
27072     10000.0         False         False     action              4  

[15211 rows x 13 columns]

In [128]:
df_achievements['release_year'] = df_achievements['release_date'].dt.year

C:\Users\Usuario\AppData\Local\Temp\ipykernel_8968\577568938.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [129]:
annual_achievements_count = df_achievements.groupby('release_year')['achievements'].sum().reset_index()

In [130]:


fig = px.bar(annual_achievements_count, 
             x='release_year', 
             y='achievements', 
             title='Evolución Anual del Número Total de Logros',
             labels={'release_year': 'Año de Lanzamiento', 'achievements': 'Total de Logros'},
             color='achievements',
             color_continuous_scale='Viridis',
             text='achievements')

fig.update_layout(
    title_font=dict(family='verdana', size=16, color='black'),
    font=dict(family='verdana', size=16),
    template='plotly_white',
    height=500,
    width=900
)

fig.show()
